bulkcopy.go

```diff
@@ -0,0 +1,576 @@
+			if b.Debug {
+				log.Printf("Adding column %s %s %#x", colname, bulkCol.ColName, bulkCol.ti.TypeId)
+			}
+		} else {
+			return fmt.Errorf("Column %s does not exist in destination table %s", colname, b.tablename)
+		}
+	}
+
+	//create the bulk command
+
+	//columns definitions
+	var col_defs string
+	for i, col := range b.bulkColumns {
+		if i != 0 {
+			col_defs += ", "
+		}
+		col_defs += "[" + col.ColName + "] " + makeDecl(col.ti)
+	}
+
+	//options
+	var with_opts []string
+
+	if b.Options.CheckConstraints {
+		with_opts = append(with_opts, "CHECK_CONSTRAINTS")
+	}
+	if b.Options.FireTriggers {
+		with_opts = append(with_opts, "FIRE_TRIGGERS")
+	}
+	if b.Options.KeepNulls {
+		with_opts = append(with_opts, "KEEP_NULLS")
+	}
+	if b.Options.KilobytesPerBatch > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("KILOBYTES_PER_BATCH = %d", b.Options.KilobytesPerBatch))
+	}
+	if b.Options.RowsPerBatch > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("ROWS_PER_BATCH = %d", b.Options.RowsPerBatch))
+	}
+	if len(b.Options.Order) > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("ORDER(%s)", strings.Join(b.Options.Order, ",")))
+	}
+	if b.Options.Tablock {
+		with_opts = append(with_opts, "TABLOCK")
+	}
+	var with_part string
+	if len(with_opts) > 0 {
+		with_part = fmt.Sprintf("WITH (%s)", strings.Join(with_opts, ","))
+	}
+
+	query := fmt.Sprintf("INSERT BULK %s (%s) %s", b.tablename, col_defs, with_part)
+
+	stmt, err := b.cn.Prepare(query)
+	if err != nil {
+		log.Fatal("Prepare failed:", err.Error())
+	}
+
+	if b.Debug {
+		log.Println(query)
+	}
+	_, err = stmt.Exec(nil)
+	if err != nil {
+		return err
+	}
+
+	b.headerSent = true
+
+	var buf = b.cn.sess.buf
+	buf.BeginPacket(packBulkLoadBCP)
+
+	//columns
+	columnMetadata := b.createColMetadata()
+	buf.Write(columnMetadata)
+
+	return nil
+}
+
+func (b *MssqlBulk) AddRow(row []interface{}) (err error) {
+
+	if b.headerSent == false {
+		err = b.sendBulkCommand()
+		if err != nil {
+			return
+		}
+	}
+
+	if len(row) != len(b.bulkColumns) {
+		return fmt.Errorf("Row does not have the same number of columns than the destination table %d %d", len(row), len(b.bulkColumns))
+	}
+
+	bytes, err := b.makeRowData(row)
+	b.cn.sess.buf.Write(bytes)
+
+	b.numRows = b.numRows + 1
+	return err
+}
+
+func (b *MssqlBulk) AddRows(rows [][]interface{}) (err error) {
+	for _, row := range rows {
+		err = b.AddRow(row)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (b *MssqlBulk) makeRowData(row []interface{}) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(tokenRow)
+
+	logcol := ""
+	for i, col := range b.bulkColumns {
+
+		if b.Debug {
+			logcol = logcol + fmt.Sprintf(" col[%d]='%v' ", i, row[i])
+		}
+		param, err := b.makeParam(row[i], col)
+		if err != nil {
+			return nil, fmt.Errorf("bulkcopy: %s", err.Error())
+		}
+
+		if col.ti.Writer == nil {
+			log.Fatalf("no writer for column: %s, TypeId: %#x", col.ColName, col.ti.TypeId)
```

Comment:
No.

This calls os.Exit, which in many applications would be a very bad thing.